### ISSUE TITLE
Add multi-factor auth (MFA) / 2-Factor Auth (2FA)

### ISSUE THREAD

**Issue by ayumi** [brave-intl/publishers#64]
Right now auth is linked to publishers' emails. If publishers emails are compromised, their Brave Publishers account is also compromised. In this case we could improve security if we offered a second factor for authentication.

@diracdeltas @mrose17

**Comment by diracdeltas** [comment]
i would highly recommend either Duo Security or Google Authenticator. Looks like for the latter you can use something like https://github.com/jaredonline/google-authenticator.

**Comment by alexwykoff** [comment]
mr : publishers.brave.com doesn't have a user / passphrase concept, we currently use a 1-time link with email round-trip. some folks have expressed concern/desire for another method.

mr : do we want to continue the rule that we don't keep secrets of any kind in the publishers.brave.com db?

bc : you can't get pwned if you don't have data to get pwned

mr : we're going to have to choose at least 1 style, @bsclifton wdyt?

bc : if we do oauth we have to keep openid, if we do 2fa we have to keep a secret, so i don't have a pref.

ay : i think we need 2fa, that way we don't use an external identity provider

bc : yeah, that solves the problem

mr : so we'll ask tilde to do what?

ay : add 2nd factor auth.  we may just have to add username & password for UX though.

mr : if possible, it'd be better to use a service so we don't need to hold TOTP
        we may need more research here

mr : will 2fa be required or optional?

ay : option is cool

bc : we need more design work around this

mr : to circle back on oauth, is there a provider we should use first?  fb or google?

ay : in that case, login/pass might just be better...

mr : maybe let's table this, we'll look into options 

to summarize : we need some design work around this @bradleyrichter @alexwykoff

**Comment by mrose17** [comment]
in reviewing this matter, my conclusion is that the initial "ask" is going to be to add YubiCloud for 2FA -- https://www.yubico.com/products/services-software/yubicloud/

this should address the concerns of folks who are security-minded, and allow us to continue to research the best approach for less-technical users.

https://github.com/titanous/yubikey appears to be the current ruby gem (though if there's another/better one, that's fine too!).

the first step is to develop the steps for the website!

**Comment by evq** [comment]
I would strongly prefer that we don't use yubicloud, it seems to only support the proprietary yubico otp.

I think we should use u2f. In addition to being the best 2fa option technically (avoids the issue of secure secret distribution and storage, key logging and phishing), it also has the advantage of being simpler for less-technical users - no qr codes to scan or codes to enter.

**Comment by ayumi** [comment]
+1 U2F. we should avoid gatekeepers and prefer free open things whenever reasonably possible.

**Comment by mrose17** [comment]
U2F is fine, per @evq the best libraries are https://github.com/castle/ruby-u2f and https://github.com/ashtuchkin/u2f ... however, we will need to have https://github.com/brave/browser-laptop/issues/518 resolved first.

since this is the last `phase-2` issue --https://github.com/brave/publishers/issues?q=is%3Aopen+is%3Aissue+label%3Aphase-2 -- the timing may work out.

**Comment by kylerschin** [comment]
@ayumi @mrose17 agreed. This is starting to bug me. I take security to "extreme" levels. Using Google Chrome to log in is not good. I would prefer brave to log in to Github or other online services.
